Add docs for external-push scaler (#193)
kedacore/keda-docs: content/docs/2.0/concepts/external-scalers.md

+++
title = "External Scalers"
weight = 500
+++

> While KEDA ships with a set of [built-in scalers](../scalers), users can also extend KEDA through a GRPC service that implements the same interface as the built-in scalers.

Reviewer: Maybe we can extend this bit with the different external scalers and how they are different before we dive in the details?

Author: I added couple more sentences. Does that match what you were thinking?

There isn't much difference between built-in and external scalers other than the responsibility of managing the GRPC server yourself.

> This document describes the external scaler interfaces and how to implement them in Go, Node, and .NET; however for more details on GRPC refer to [the official GRPC documentation](https://grpc.io/docs/)

Reviewer: Add newline to optimize the rendered output
```suggestion

This document describes the external scaler interfaces and how to implement them in Go, Node, and .NET; however for more details on GRPC refer to [the official GRPC documentation](https://grpc.io/docs/)
```

> 

Reviewer: ```suggestion

Want to learn about existing external scalers? Explore our [external scaler community](https://github.com/kedacore/external-scalers).

```

## Overview

### Built-in scalers interface

Built-in scalers implement one of the following go interfaces:

```go
type Scaler interface {
	GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error)
	GetMetricSpecForScaling() []v2beta2.MetricSpec
	IsActive(ctx context.Context) (bool, error)
	Close() error
}

type PushScaler interface {
	Scaler

	Run(ctx context.Context, active chan<- bool)
}
```

The `Scaler` interface defines 4 methods:

> - `GetMetrics` and `GetMetricsSpecForScaling` that are used to build the HPA definition for that particular scaler.

Reviewer: What should we expect here? It's not fully clear to me.

Author: There is a bit more details about them in the "Implementing" sections 

> `GetMetricSpec` returns the target value for the HPA definition for the scaler. This scaler will define a static target of 10, but the threshold value is often specified in the metadata for other scalers.

https://github.com/ahmelsayed/keda-docs/blob/push-scaler/content/docs/2.0/concepts/external-scalers.md#6-implementing-getmetrics

> `GetMetrics` returns the value of the metric refered to from `GetMetricSpec`, in this example it's `earthquakeThreshold`. Refer to the HPA docs for how HPA calculates `replicaCount` based on metric value and target value.

https://github.com/ahmelsayed/keda-docs/blob/push-scaler/content/docs/2.0/concepts/external-scalers.md#6-implementing-getmetrics

It's easier to explain with an example, but I can expand the description there a bit. how about

> `GetMetricSpec` returns the target value for the HPA definition for the scaler. For more details refer to [Implementing `GetMetricSpec`]()

>  `GetMetrics` returns the value of the metric refered to from `GetMetricSpec`. For more details refer to [Implementing `GetMetrics`]()

?

Reviewer: It helps a lot, thanks!

Examples are always welcome but not a must!

- `IsActive` is called on `pollingInterval` defined in the ScaledObject/ScaledJob CRDs and scaling to 1 happens if this returns true
- `Close` is called to allow the scaler to clean up connections or other resources.

> The `PushScaler` interface adds `Run` method. The `Run` method recieves a push channel (`active`), that the scaler can push `true` to any time to force activation independently from `pollingInterval`.

Reviewer: Activation == scaling action or not?
```suggestion
The `PushScaler` interface adds `Run` method. The `Run` method receives a push channel (`active`), that the scaler can push `true` to any time to force scaling action independently from `pollingInterval`.
```

### External Scaler GRPC interface

> KEDA comes with 2 external scalers [`external`](../scalers/external.md) and [`external-push`](../scalers/external-push.md). The configuration in the ScaledObject points to a GRPC service endpoint that implements the following GRPC contract [`externalscaler.proto`](https://github.com/kedacore/keda/blob/v2/pkg/scalers/externalscaler/externalscaler.proto):

Reviewer: Optimizing rendered output
```suggestion
KEDA comes with 2 external scalers [`external`](../scalers/external.md) and [`external-push`](../scalers/external-push.md).

The configuration in the ScaledObject points to a GRPC service endpoint that implements the following GRPC contract [`externalscaler.proto`](https://github.com/kedacore/keda/blob/v2/pkg/scalers/externalscaler/externalscaler.proto):
```

```xml
  <ItemGroup>
    <Protobuf Include="Protos\externalscaler.proto" GrpcServices="Server" />
  </ItemGroup>
```

{{< /collapsible >}}

{{< collapsible "Javascript" >}}
2.1. Prepare the project

```bash
npm install --save grpc request
```
{{< /collapsible >}}

<br />

#### 3. Implementing `IsActive`

Just like `IsActive(ctx context.Context) (bool, error)` in the go interface, `IsActive` method in the GRPC interface is called every `pollingInterval` with a `ScaledObjectRef` object that contains the scaledObject name, namespace, and scaler metadata. This section implements an external scaler that queries earthquakes from https://earthquake.usgs.gov/ and scales the deployment if there has been more than 2 earthquakes with `magnitude > 1.0` around a particular longitude/latitude in the previous day

`ScaledObject`

```yaml
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: scaledobject-name
  namespace: scaledobject-namespace
spec:
  scaleTargetRef:
    name: deployment-name
  triggers:
    - type: external
      metadata:
        scalerAddress: earthquake-scaler:9090
        longitude: "-122.335167"
        latitude: "47.608013"
```

{{< collapsible "Golang" >}}

> Full implementation can be found here: https://github.com/ahmelsayed/external-scaler-samples

Reviewer: Can we donate this to `kedacore`?

Author: yes, was just preparing the repo. Will move it there.

Reviewer: Thanks and also for the repo, very useful to have a reference as well!

        url: `https://earthquake.usgs.gov/fdsnws/event/1/query?${query}`,
        json: true,
      }, (err, resp, data) => {
        if (err) {
          callback({
            code: grpc.status.INTERNAL,
            details: err,
          })
        } else if (resp.statusCode !== 200) {
          callback({
            code: grpc.status.INTERNAL,
            details: `expected status 200, got ${resp.statusCode}`
          })
        } else {
          // count how many earthquakes with mag > 1.0
          let count = 0
          data.features.forEach(i => {
            if (i.properties.mag > 1.0) {
              count++
            }
          })
          callback(null, {
            result: count > 2,
          })
        }
      })
    }
  }
})

server.bind('127.0.0.1:9090', grpc.ServerCredentials.createInsecure())
console.log('Server listening on 127.0.0.1:9090')

server.start()
```
{{< /collapsible >}}

<br />

#### 4. Implementing `StreamingIsActive`

Unlike `IsActive`, `StreamingIsActive` is called once when KEDA reconciles the `ScaledObject`, and expects the external scaler to push `IsActiveResponse` whenever the scaler needs KEDA to activate the deployment.

> This implementation creates a timer and queries USGS APIs on that timer, effictivly ignoring `pollingInterval` set in the scaledObject. Alternativly any other asynchronus event can be used instead of a timer, like an HTTP request, or a network connection.

Reviewer: ```suggestion
This implementation creates a timer and queries USGS APIs on that timer, effectively ignoring `pollingInterval` set in the scaledObject. Alternatively any other asynchronous event can be used instead of a timer, like an HTTP request, or a network connection.
```

{{< collapsible "Golang" >}}
```golang
func (e *ExternalScaler) GetMetricSpec(context.Context, *pb.ScaledObjectRef) (*pb.GetMetricSpecResponse, error) {
	return &pb.GetMetricSpecResponse{
		MetricSpecs: []*pb.MetricSpec{{
			MetricName: "earthquakeThreshold",
			TargetSize: 10,
		}},
	}, nil
}
```
{{< /collapsible >}}

{{< collapsible "C#" >}}
```csharp
public override Task<GetMetricSpecResponse> GetMetricSpec(ScaledObjectRef request, ServerCallContext context)
{
  var resp = new GetMetricSpecResponse();

  resp.MetricSpecs.Add(new MetricSpec
  {
    MetricName = "earthquakeThreshold",
    TargetSize = 10
  });

  return Task.FromResult(resp);
}
```
{{< /collapsible >}}

{{< collapsible "Javascript" >}}
```js
server.addService(externalScalerProto.externalscaler.ExternalScaler.service, {
  // ...
  getMetricSpec: (call, callback) => {
    callback(null, {
      metricSpecs: [{
        metricName: 'earthquakeThreshold',
        targetSize: 10,
      }]
    })
  }
})
```
{{< /collapsible >}}

<br />

#### 6. Implementing `GetMetrics`

> `GetMetrics` returns the value of the metric refered to from `GetMetricSpec`, in this example it's `earthquakeThreshold`. Refer to the [HPA docs](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/) for how HPA calculates replicaCount based on metric value and target value.

Reviewer: ```suggestion
`GetMetrics` returns the value of the metric refered to from `GetMetricSpec`, in this example it's `earthquakeThreshold`.

Refer to the [HPA docs](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/) for how HPA calculates replicaCount based on metric value and target value.
```

I would even maybe move the last part underneight the samples as it's less relevant here

+++
title = "External Scalers"
weight = 500
+++

> While KEDA ships with a set of [built-in scalers](../scalers), users can also extend KEDA through a GRPC service that implements the same interface as the built-in scalers. Built-in scalers run in the KEDA process/pod, while external scalers require an externally managed GRPC server that's accessible from KEDA with optional [TLS authentication](https://grpc.io/docs/guides/auth/). KEDA itself acts as a GRPC client and it exposes similar service interface for the built-in scalers, so external scalers can fully replace built-in ones.

Reviewer: ```suggestion
While KEDA ships with a set of [built-in scalers](../scalers), users can also extend KEDA through a GRPC service that implements the same interface as the built-in scalers.

Built-in scalers run in the KEDA process/pod, while external scalers require an externally managed GRPC server that's accessible from KEDA with optional [TLS authentication](https://grpc.io/docs/guides/auth/). KEDA itself acts as a GRPC client and it exposes similar service interface for the built-in scalers, so external scalers can fully replace built-in ones.
```
